memgpt/agent.py

```diff
@@ -320,6 +292,65 @@ def messages(self) -> List[dict]:
     def messages(self, value):
         raise Exception("Modifying message list directly not allowed")
 
+    def link_tools(self, tools: List[Tool]):
+        """Bind a tool object (schema + python function) to the agent object"""
+
+        # tools
+        for tool in tools:
+            assert tool, f"Tool is None - must be error in querying tool from DB"
+            assert tool.name in self.agent_state.tools, f"Tool {tool} not found in agent_state.tools"
+        for tool_name in self.agent_state.tools:
+            assert tool_name in [tool.name for tool in tools], f"Tool name {tool_name} not included in agent tool list"
+
+        # Store the functions schemas (this is passed as an argument to ChatCompletion)
+        self.functions = []
+        self.functions_python = {}
+        env = {}
+        env.update(globals())
+        for tool in tools:
+            # WARNING: name may not be consistent?
+            if tool.module:  # execute the whole module
+                exec(tool.module, env)
+            else:
+                exec(tool.source_code, env)
+            self.functions_python[tool.name] = env[tool.name]
+            self.functions.append(tool.json_schema)
+        assert all([callable(f) for k, f in self.functions_python.items()]), self.functions_python
+
+    def _load_messages_from_recall(self, message_ids: List[str]) -> List[Message]:
+        """Load a list of messages from recall storage"""
+
+        # Pull the message objects from the database
+        message_objs = [self.persistence_manager.recall_memory.storage.get(msg_id) for msg_id in message_ids]
+        assert all([isinstance(msg, Message) for msg in message_objs])
+
+        return message_objs
+
+    def _validate_message_buffer_is_utc(self):
+        """Iterate over the message buffer and force all messages to be UTC stamped"""
+
+        for m in self._messages:
+            # assert is_utc_datetime(m.created_at), f"created_at on message for agent {self.agent_state.name} isn't UTC:\n{vars(m)}"
+            # TODO eventually do casting via an edit_message function
+            if not is_utc_datetime(m.created_at):
+                printd(f"Warning - created_at on message for agent {self.agent_state.name} isn't UTC (text='{m.text}')")
+                m.created_at = m.created_at.replace(tzinfo=datetime.timezone.utc)
+
+    def set_message_buffer(self, message_ids: List[str], force_utc: bool = True):
+        """Set the messages in the buffer to the message IDs list"""
+
+        message_objs = self._load_messages_from_recall(message_ids=message_ids)
+
+        # set the objects in the buffer
+        self._messages = message_objs
+
+        # bugfix for old agents that may not have had UTC specified in their timestamps
+        if force_utc:
+            self._validate_message_buffer_is_utc()
+
+        # also sync the message IDs attribute
+        self.agent_state.message_ids = message_ids
```

Comment:
@sarahwooders do we need to do this?

Or can we assume that every time an external function accesses/reads `agent.agent_state.message_ids`, it will do an `agent.update_state` first?

Because there are a few other places where `agent._messages` gets modified but we don't sync `agent.agent_state.message_ids`.

If we care about this being in sync, we may want to turn `agent.agent_state` into a getter property that automatically refreshes/builds `message_ids` from `_messages` (or we need to eagerly update `self.agent_state.message_ids` whenever we modify `self._messages`.